### PR TITLE
AI Target Permanence

### PR DESCRIPTION
**Pull request by SuicidalInsanity**
Improvements
-AI will now head to target's last known position if contact lost (out of sight range/angle/radar detection), with accuracy of predicted position decaying over time.
-Adds Proc Wing/ProcPart HP cap setting to clamp proc HP to a configurable maximum.
-AI will now enable radars when guard mode enabled.
Misc Fixes
-Fix target A2A/A2G preference GUI labels
-Fix some EMP weapon Selection logic
-Fix AI withdrawal after firing nukes/EMPs at close range

### CODE REVIEW

File: BDArmory/Control/BDModulePilotAI.cs [hunk 1]
```diff
@@ -1321,14 +1321,30 @@ IEnumerator setVar(string name, float value)
...
+            if (weaponManager.detectedTargetTimeout > weaponManager.targetScanInterval)
+            {
+                timer += Time.fixedDeltaTime;
+                if (timer >= 50) //add some error to the predicted position every second
+                {
+                    Vector3 staleTarget = new Vector3();
+                    staleTarget.x = UnityEngine.Random.Range(0f, (float)lastKnownVector.magnitude / 10);
+                    staleTarget.y = UnityEngine.Random.Range(0f, (float)lastKnownVector.magnitude / 10);
+                    staleTarget.z = UnityEngine.Random.Range(0f, (float)lastKnownVector.magnitude / 20);
+                    timer = 0;
```
Comment (SuicidalInsanity) [inline BDArmory/Control/BDModulePilotAI.cs:1341]: Could perhaps play around with the values for more/less target position decay; +/- ~20-30m/s isn't all that much, even over 30 seconds

File: BDArmory/UI/BDArmorySetup.cs [hunk 2]
```diff
@@ -1645,11 +1645,11 @@ void WindowBDAToolbar(int windowID)
                     }
                     else
                     {
-                        textNumFields["targetWeightRange"].tryParseValue(GUI.TextField(new Rect(leftIndent + (90), (priorityLines * entryHeight), contentWidth - 90 - 38, entryHeight), textNumFields["targetWeightRange"].possibleValue, 4));
-                        ActiveWeaponManager.targetWeightAirPreference = (float)textNumFields["targetWeightRange"].currentValue;
+                        textNumFields["targetWeightAirPreference"].tryParseValue(GUI.TextField(new Rect(leftIndent + (90), (priorityLines * entryHeight), contentWidth - 90 - 38, entryHeight), textNumFields["targetWeightAirPreference"].possibleValue, 4));
```
Comment (BrettRyland) [inline BDArmory/UI/BDArmorySetup.cs:1648]: This is missing a
```
{ "targetWeightAirPreference", gameObject.AddComponent<NumericInputField>().Initialise(0, ActiveWeaponManager.targetWeightAirPreference, -10, 10) },
```
with appropriate limits on line 783.

File: BDArmory/Control/BDModulePilotAI.cs [hunk 3]
```diff
@@ -1673,6 +1689,8 @@ bool RamTarget(FlightCtrlState s, Vessel v)
             return true;
         }
 
+        Vector3 staleTarget = Vector3.zero;
```
Comment (BrettRyland) [inline BDArmory/Control/BDModulePilotAI.cs:1692]: Are these position and velocity of the stale target? I'd suggest renaming them `staleTargetPosition` and `staleTargetVelocity` so they're self-explanatory.

File: BDArmory/Control/BDModulePilotAI.cs [hunk 4]
```diff
@@ -1321,14 +1321,30 @@ IEnumerator setVar(string name, float value)
             yield return new WaitForFixedUpdate();
             typeof(BDModulePilotAI).GetField(name).SetValue(this, value);
         }
-
+        float timer = 0;
```
Comment (BrettRyland) [inline BDArmory/Control/BDModulePilotAI.cs:1324]: Can we use a better name for this, `staleTargetTimer` for example?

**Pull request merged**